feat(store): Moved connection logic to a StoreBackend service

MikeRyanDev opened the pull request:
This is a pretty heavy refactor. Changes:
1. Introduce a `StoreBackend` service. This change makes it easy for future devtools to lift both the state and action stream without having to modify the `Store` class. This service has three responsibilities:
   1. Apply middleware to the action dispatcher
   2. Connect the action dispatcher to any subject
   3. Replace the action reducer during runtime
2. Re-introduce the `Dispatcher` service. This makes it easy to apply middleware to the action stream and for devtools to wrap it.
3. Create a pipeline for applying a middleware function to the `Action` stream. This uses the `let` operator on the `Action` observable.
4. Simplify `reducer` throughout entire library. Now reducers are just a function. Simplifies `replaceReducer` as well.
5. Move all Angular 2 specific code to `ng2.ts`. This just wires up the services using Angular 2's DI. `provideStore(reducerFn, initialState)` sets up the services, and `useMiddleware(...middleware: Middleware[])` adds middleware to the middleware pipeline. This should make it a little easier for others to use ngrx/store with a different framework.  

I've modified all of the tests to verify they work against the new architecture. Realizing now that I'm missing tests for middleware, those are forthcoming.

robwormald commented on spec/store_spec.ts:
```diff
@@ -180,13 +189,13 @@ describe('ngRx Store', () => {
       expect(currentState).toEqual({counter1: 1, counter2: 2, counter3: 1});
 
-      store.replaceReducers({'dynamicCounter' : counterReducer});
+      backend.replaceReducer(combineReducers({ counter1: counterReducer, dynamicCounter: counterReducer}));
```
curious as to why this isn't tested against the store's replaceReducer API?

robwormald commented:
let's move all the ng2 specific stuff to the index.ts file, if possible, and make that the main entry point (make sure the `typings` field matches in package.json) - if we decide to ship an ng1 ready release we can add a new entry point for that, and generate a different package.json

robwormald commented:
while you're writing tests, could you add a couple to test the behavior of what happens if an error gets nexted into the store and the dispatcher? we may want to noop those out.

robwormald commented on src/store.ts:
```diff
@@ -1,54 +1,23 @@
+  ) {
+    super(initialState);
 
-  next(action: any) {
```
next should be left in here, and now you've reintroduced the dispatcher, redirected through that. Store should still implement the `Observer` interface, and w/o overriding next, it would simply set the state and short circuit the whole process.

robwormald replied:
eg somebody should be able to do `actions$.subscribe(store)`

MikeRyanDev replied:
Added it back in. Changed signature of `backend.connect` to take a function instead of an observer, letting me safely override `next`

MikeRyanDev commented:
Updated: 
1. Re-exported `ng2.ts` in `index.ts`. Updated `package.json` to point to `./dist/index.ts` for the main and `./dist/index.d.ts` for the typings main
2. Got the `npm run lint` command working and fixed all of the linting errors
3. Restored observer capability for `Store` class. Added a unit test to verify.

I'll look into error handling next.

robwormald commented:
looking good. let me know when you're ready for a final review and we'll merge it in.

MikeRyanDev committed:
feat(StoreBackend): Refactored lib to use a backend service for middleware and devtools

MikeRyanDev commented:
Ok, changed again:
1. Changed `Store` to forward errors to `Dispatcher`. Should allow for a followup PR to address error handling, though you could now pretty easily write an error handling middleware.
2. Added some tests for the `StoreBackend` service.

I think this is ready for a final review.
